src/meepgeom.cpp

```diff
@@ -2436,61 +2450,64 @@ void material_grids_addgradient_point(meep::realnum *v, std::complex<double> *fi
-
-  // size indices
-  vector3 max_corner = vec_to_vector3(where.get_max_corner());
-  vector3 min_corner = vec_to_vector3(where.get_min_corner());
-  s1 = (vec_to_vector3(min_max_loc[1]).x - vec_to_vector3(min_max_loc[0]).x) / n1;
-  s2 = (vec_to_vector3(min_max_loc[1]).y - vec_to_vector3(min_max_loc[0]).y) / n2;
-  s3 = (vec_to_vector3(min_max_loc[1]).z - vec_to_vector3(min_max_loc[0]).z) / n3;
-  // starting point
-  c1 = n1 <= 1 ? 0 : vec_to_vector3(min_max_loc[0]).x;
-  c2 = n2 <= 1 ? 0 : vec_to_vector3(min_max_loc[0]).y;
-  c3 = n3 <= 1 ? 0 : vec_to_vector3(min_max_loc[0]).z;
-
-  // Loop over x,y,z and frequency dimensions
+  meep::vec min_max_loc[2] = {meep::vec(0,0,0),meep::vec(0,0,0)}; // extremal points in subgrid
+  bool collapse = false, snap = true;
+  meep::component field_dir[3] = {meep::Ex, meep::Ey, meep::Ez};
+  size_t dims[9] = {1, 1, 1, 1, 1, 1, 1, 1, 1};
+  for (int c = 0; c < 3; c++) {
+
+    int rank = f->get_array_slice_dimensions(where, &dims[3 * c], dirs, collapse, snap, min_max_loc,
+                                             0, field_dir[c]);
+    
+    vector3 max_corner = vec_to_vector3(min_max_loc[1]);
+    meep::realnum max_c_array[3] = {max_corner.x, max_corner.y, max_corner.z};
+    vector3 min_corner = vec_to_vector3(min_max_loc[0]);
+    meep::realnum min_c_array[3] = {min_corner.x, min_corner.y, min_corner.z};
+    for (int ci = 0; ci < 3; ci++) {
+      s[c][ci] = (max_c_array[ci] - min_c_array[ci]) == 0 ? 0 : (max_c_array[ci] - min_c_array[ci]) / (dims[3 * c + ci] - 1);
+      cen[c][ci] = dims[3 * c + ci] <= 1 ? 0 : min_c_array[ci];
+    }
+  }
+
+  // Loop over component, x, y, z, and frequency dimensions
   // TODO speed up with MPI (if needed)
-  for (int i1 = 0; i1 < n1; ++i1) {
-    for (int i2 = 0; i2 < n2; ++i2) {
-      for (int i3 = 0; i3 < n3; ++i3) {
-        for (int i4 = 0; i4 < n4; ++i4) {
-          int xyz_index = (((i1 * n2 + i2) * n3 + i3) * n4 + i4) * 3;
-          std::complex<double> *fields_a_cur, *fields_f_cur;
-          fields_a_cur = &fields_a[xyz_index];
-          fields_f_cur = &fields_f[xyz_index];
-          p.x = i1 * s1 + c1;
-          p.y = i2 * s2 + c2;
-          p.z = i3 * s3 + c3;
-          material_grids_addgradient_point(v, fields_a_cur, fields_f_cur, p, scalegrad,
-                                           frequencies[i4], geometry_tree);
+  size_t c_offset = 0;
+  for (int c = 0; c < 3; c++) {             // component
+    n2 = dims[c * 3]; n3 = dims[c * 3 + 1]; n4 = dims[c * 3 + 2];
+    c1 = cen[c][0]; c2 = cen[c][1]; c3 = cen[c][2];
+    s1 = s[c][0]; s2 = s[c][1]; s3 = s[c][2];
+
+    for (int i1 = 0; i1 < nf; ++i1) {       // freq
+      meep::realnum *v_cur = &v[ng * i1];
+      for (int i2 = 0; i2 < n2; ++i2) {     // x
+        for (int i3 = 0; i3 < n3; ++i3) {   // y
+          for (int i4 = 0; i4 < n4; ++i4) { // z
+            int xyz_index = c_offset + (((i1 * n2 + i2) * n3 + i3) * n4 + i4);
```

Comment:
Might be clearer to do 
```c
int xyz_index = ((((c * nf + i1) * n2 + i2) * n3 + i3) * n4 + i4);
```
here rather than computing `c_offset` separately.

Reply:
Nevermind, this won't work since each component has different-sized grids.